static/src/javascripts/projects/common/utils/steady-page.js

```diff
@@ -0,0 +1,187 @@
+        var batchHeightsBeforeInsert;
+
+        promise = fastdom.read(function(){
+            while (!q.empty()) {
+                // Take the current queue items and add them to the batch array
+                batch.push(q.dequeue());
+            }
+
+            return batch;
+        })
+            .then(function(){
+                return getHeightOfAllContainers(batch);
+            })
+            .then(function(heightsBeforeIns) {
+                batchHeightsBeforeInsert = heightsBeforeIns || 0;
+                return insertElements(batch);
+            })
+            .then(function(){
+                return getHeightOfAllContainers(batch).then(function(newHeights) {
+                    return assign(state, {
+                        newHeight: newHeights - batchHeightsBeforeInsert
+                    });
+                });
+            })
+            .then(function(state){
+                if (q.empty()) {
+                    // If the queue is empty (no more elements need to be added to the page) we immediately scroll
+                    return state.newHeight + state.prevHeight + state.scrollY;
+                } else {
+                    // If there are elements waiting to be added to the page we take the previous container's heights
+                    // and recursively call the function so that we only scroll the page once the queue is empty -
+                    // this prevents excessive and jarring scrolling
+                    return go(assign(state, {
+                        prevHeight: state.prevHeight + state.newHeight
+                    }));
+                }
+            });
+
+        return promise;
+
+    }
+
+    /**
+     * Insert an element into the page
+     * Use if your element doesn't exist and is inserted into a container
+     * ** Don't use fastdom - it is handled in this utility **
+     * @param {HTMLElement} container The element that the component is being inserted into
+     * @param {Function} cb Should contain all functionality that displays and lays-out the element
+     */
+    function insert(container, cb) {
+        if (!config.switches.steadyPageUtil) {
+            return fastdom.write(cb);
+        }
+
+        var initialState = {
+            scrollY: window.scrollY,
+            prevHeight: 0
+        };
+
+        q.enqueue({
+            container: container,
+            cb: cb
+        });
+
+        return (running ? promise : go(initialState)).then(scrollThePage);
+    }
+
+
+    /**
+     * Given a batch, call all of the callbacks on the insertion object
+     *
+     * @param  {Array} batch
+     */
+    function insertElements (batch) {
+        return fastdom.write(function () {
+            batch.forEach(function (insertion) {
+                insertion.cb();
+            });
+        });
+    }
+
+    function scrollThePage (scrollY) {
+        if (scrollY) {
+            window.scrollTo(0, scrollY);
+            running = false;
```

Comment:
I think this should be moved into the `go` routine, since there is a chance a task might be queued *after* the go routine has finished and *before* this callback is called, and then it would never get picked up. What do you think?